### PR TITLE
SafeRenderManager retention workaround

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
SafeRenderManager gets retained by the rendered SafeRenderManagerHelper, which in some cases was being retained by React, seemingly because of the following behavior:

_From https://github.com/facebook/react/issues/16087:_
> Props/child trees retained by alternate children. Similarly, children that was just removed can sometimes between retained by the alternate copy of that. That is until that node gets another update on it which clears out the old children. These cases are fairly unusual and fix themselves eventually as the app lives on.

These retentions eventually resolve themselves, so they're not true leaks, but they can show up as leaks in memory leak tests that look for disposed-but-retained objects (via Disposable's LeakFlag mechanism).

There are cases where it's very difficult to avoid these kinds of React retention paths. However, this case happens to be easier to avoid, and doing so would reduce noise and some false positives in these memory leak tests, so we want to update SafeRenderManager to not be retained by the tree it renders.

This won't prevent React from retaining the React tree it renders, which may retain other Disposable instances and still be flagged as leaks; only the SafeRenderManager object itself will no longer be retained. However, it's possible that the way content is "safely" unmounted (by rerendering SafeRenderManagerHelper with empty children before unmounting) helps prevent consumer-provided content being retained, but we're not sure at this time. 

## Changes
- Prevent SafeRenderManager from being retained by the React tree it renders. The retainers were two callback refs that directly retained SafeRenderManager due to being tearoffs:
    - `_contentCallbackRef` - fixed by using a ref object instead
    - `_helperRef` - fixed by instead retaining a `_selfRef` ref object which holds a SafeRenderManager, and gets cleared out on disposal
- Add additional test coverage to ensure callback chaining works as intended with different component types

Note that callback refs aren't inherently problematic, nor was how SafeRenderManager was being retained by the tree it rendered. The root issue here is React retaining things longer than we'd like/expect it to, and this issue helps work around that to reduce noise in tests. 

#### Release Notes
- Update `SafeRenderManager` to not be retained by the React tree it renders, to aid in memory leak testing. Outside of that context, this change has negligible effects.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/util/safe_render_manager/safe_render_manager_helper.dart
Comment (greglittlefield-wf) [inline lib/src/util/safe_render_manager/safe_render_manager_helper.dart:70]: I upgraded the boilerplate while I was in here, but this should be fine this component is completely internal

**Review by danielway-wk — APPROVED** [review]
I've confirmed this resolves the specific leak I was investigating for spreadsheets, and it appears that several other tests were also impacted by this issue and are resolved by this fix (still waiting on a final test run).

File: lib/src/util/safe_render_manager/safe_render_manager.dart
Comment (greglittlefield-wf) [inline lib/src/util/safe_render_manager/safe_render_manager.dart:61]: This is technically a breaking change since it removes a setter, but setting `contentRef` shouldn't have ever been supported, and I can't find any code that sets this.

Comment (greglittlefield-wf) [reply lib/src/util/safe_render_manager/safe_render_manager.dart:61]: Actually, I decided to avoid a breakage here ee8ab08a8c48308b9218605ce62c1b15544b66c3

**Review by keatoncarter-wf — APPROVED** [review]

**Review by gregthompson-wk — APPROVED** [review]

**Review by jediaelfraser-wk — APPROVED** [review]

**Review by danielway-wk — APPROVED** [review]

**Review by aaronlademann-wf — APPROVED** [review]
+1 / QA +1 (CI passing)

**Comment by danielway-wk** [comment]
FYI Luke is QAing in DPC before we merge.

**Comment by lukeanderson-wk** [comment]
> FYI Luke is QAing in DPC before we merge.

Yep waiting on https://ci.webfilings.com/build/3406321

**Comment by lukeanderson-wk** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
